manpages/man/tiledb_update_bin.1
.TH tiledb_update_bin 1 "29 June 2015" "Version 0.1" "TileDB programs"
 
.SH NAME
tiledb_update_bin - loads a collection of binary files into an array,
which represents updates to the existing data of the array.

.SH SYNOPSIS
.B tiledb_update_bin 
.BI "-w " "directory " "-A " "array_name " "-p " "path" 
[
.BI "-m " "mode"
]

.SH DESCRIPTION
.B tiledb_update_bin
This is very similar to \fBtiledb_load_bin\fR(1). The difference is that the 
loaded data create a new fragment in the array, which may be consolidated with
some or all of the existing ones. The update mechanism is described in detail
in the manpages of \fBtiledb_define_array\fR(1).

The following conventions apply to the \fBSYNOPSIS\fR section:

.TS
tab (@);
c lx .
\fBbold text\fR @ type exactly as shown
\fIitalic text\fR @ replace with appropriate argument
[\fB\-a \fIarg\fR]@ any or all options within [ ] are optional
.TE

.SH OPTIONS
.TP
.BI "-w" " directory" "\fR, " \fB --workspace=\fIdirectory\fR  
The path to the workspace folder where the TileDB will be stored. The path
must exist, and the caller must have read and write permissions on it.

.TP
.BI "-A" " array_name" "\fR, " \fB --array-name=\fIarray_name\fR  
The name of the array where the data will be loaded. The array
must be already defined. 

.TP
.BI "-p" " path" "\fR, " \fB --path=\fIpath\fR  
The path to a binary file or to a directory of binary files. If it is
a file, then this single file will be loaded. If it is a directory,
\fBall\fR the files in the directory will be loaded.

.TP
.BI "-m" " mode" "\fR, " \fB --mode=\fImode\fR  
It can be either \fBsorted\fR or \fBunsorted\fR. This is also optional;
if it is omitted, the default is \fBunsorted\fR. This indicates whether the
cells in the binary file(s) are sorted along the cell order defined in
the array schema. This choice will have a great effect on performance
(sorted cells are loaded substantially faster).

.SH EXAMPLES
.TP
\fBtiledb_update_bin\fR \\ 
    \fB-w \fRmy_workspace/ \\
    \fB-A \fRmy_array \\
    \fB-p \fRmy_array.bin

This updates array \fImy_array\fR in workspace folder \fImy_workspace\fR with 
the data stored in the binary file \fImy_array.bin\fR. It is assumed that the 
cells are unsorted.

.TP
\fBtiledb_update_bin\fR \\ 
    \fB-w \fRmy_workspace/ \\
    \fB-A \fRmy_array \\
    \fB-p \fRmy_bin_files/

This is the same as the first example, but now all the files included in
directory \fImy_bin_files/\fR will be loaded into the array.

.TP
\fBtiledb_update_bin\fR \\ 
    \fB-w \fRmy_workspace/ \\
    \fB-A \fRmy_array \\
    \fB-p \fRmy_bin_files/ \\
    \fB-m \fRsorted

This is the same as the second example, but now the cells in each file
(but not necessarily across files) are sorted on the cell order defined
in the array schema.

.SH EXIT STATUS
.TP 
.B 0
Successful execution
.TP 
.B -1
Parser error
.TP 
.B -2
Undefined array
.TP 
.B -3
File operation failed
.TP 
.B -7
TileDB initialization failed
.TP 
.B -8
TileDB finalization failed
.TP 
.B -14
Failed to open array
.TP 
.B -15
Failed to close array

.SH SEE ALSO
.BR "tiledb_define_array" "(1), " "tiledb_load_bin" "(1), " \
"tiledb_update_csv" "(1), " \
"tiledb_export_csv" "(1), " "tiledb_generate_data" "(1) "

.SH AUTHOR
Stavros Papadopoulos <http://people.csail.mit.edu/stavrosp>, Copyright (c) 2015


tiledb_cmd/src/tiledb_update_bin.cc
/**
 * @file   tiledb_update_bin.cc
 * @author Stavros Papadopoulos <stavrosp@csail.mit.edu>
 *
 * @section LICENSE
 *
 * The MIT License
 * 
 * @copyright Copyright (c) 2014 Stavros Papadopoulos <stavrosp@csail.mit.edu>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * 
 * @section DESCRIPTION
 *
 * Implements command "tiledb_update_bin".
 */

#include "csv_line.h"
#include "special_values.h"
#include "tiledb.h"
#include <getopt.h>
#include <iostream>
#include <string>

/** 
 * Parses the command options. It returns 0 on success. On error, it prints
 * a message on stderr and returns -1.
 */                                                            
int parse_options(
    int argc, 
    char** argv, 
    std::string& workspace,
    std::string& array_name, 
    std::string& path, 
    bool& sorted) {
  // Initialization
  workspace = "";
  array_name = "";
  path = "";
  std::string mode("");

  // Define long options
  static struct option long_options[] = 
  {
    {"array-name",1,0,'A'},
    {"mode",1,0,'m'},
    {"path",1,0,'p'},
    {"workspace",1,0,'w'},
    {0,0,0,0},
  };
  // Define short options
  const char* short_options = "A:m:p:w:";
  // Get options
  int c;
  int option_num = 0;
  while((c=getopt_long(argc, argv, short_options, long_options, NULL)) >= 0) {
    ++option_num;
    switch(c) {
      case 'A':
        if(array_name != "") {
          std::cerr << ERROR_MSG_HEADER 
                    << " More than one array names provided.\n";
          return -1;
        }
        array_name = optarg;
        break;
      case 'm':
        if(mode != "") {
          std::cerr << ERROR_MSG_HEADER 
                    << " More than one modes provided.\n";
          return -1;
        }
        mode = optarg;
        break;
      case 'p':
        if(path != "") {
          std::cerr << ERROR_MSG_HEADER 
                    << " More than one paths provided.\n";
          return -1;
        }
        path = optarg;
        break;
      case 'w':
        if(workspace != "") {
          std::cerr << ERROR_MSG_HEADER 
                    << " More than one workspaces provided.\n";
          return -1;
        }
        workspace = optarg;
        break;
      default:
        return -1;
    }
  }

  // Check number of arguments
  if((argc-1) != 2*option_num) {
    std::cerr << ERROR_MSG_HEADER << " Arguments-options mismatch.\n";
    return -1;
  }
  // Check if correct arguments have been ginen
  if(array_name == "") {
    std::cerr << ERROR_MSG_HEADER << " Array name not provided.\n";
    return -1;
  }
  if(path == "") {
    std::cerr << ERROR_MSG_HEADER << " Path not provided.\n";
    return -1;
  }
  if(workspace == "") {
    std::cerr << ERROR_MSG_HEADER << " Workspace not provided.\n";
    return -1;
  }

  // Check number of arguments
  // ----- array name
  CSVLine temp;
  temp << array_name;
  if(temp.size() > 1) {
    std::cerr << ERROR_MSG_HEADER 
              << " More than one array names provided.\n";
    return -1;
  }
  array_name = temp.str(); // Ignore trailing comma
  // ----- mode
  temp.clear();
  temp << mode;
  if(temp.size() > 1) {
    std::cerr << ERROR_MSG_HEADER 
              << " More than one modes provided.\n";
    return -1;
  }
  mode = temp.str(); // Ignore trailing comma
  // ----- path
  temp.clear();
  temp << path;
  if(temp.size() > 1) {
    std::cerr << ERROR_MSG_HEADER 
              << " More than one paths provided.\n";
    return -1;
  }
  path = temp.str(); // Ignore trailing comma
  // ----- workspace
  temp.clear();
  temp << workspace;
  if(temp.size() > 1) {
    std::cerr << ERROR_MSG_HEADER 
              << " More than one workspaces provided.\n";
    return -1;
  }
  workspace = temp.str(); // Ignore trailing comma

  // Get sorted
  if(mode == "" || mode == "unsorted") {
    sorted = false;
  } else if(mode == "sorted") {
    sorted = true;
  } else {
    std::cerr << ERROR_MSG_HEADER 
              << " Unknown mode.\n";
    return -1;
  }

  return 0;
}

int main(int argc, char** argv) {
  // Arguments
  std::string workspace;
  std::string array_name;
  std::string path;
  bool sorted;

  // Stores the return code of the various functions below
  int rc; 
 
  // Parse command line
  rc = parse_options(argc, argv, workspace, array_name, path, sorted);
  if(rc) {
    std::cerr << ERROR_MSG_HEADER << " Failed to parse the command line.\n";
    return TILEDB_EPARSE;
  }

  // Initialize TileDB
  TileDB_CTX* tiledb_ctx;
  rc = tiledb_init(workspace.c_str(), tiledb_ctx);
  if(rc) {
    std::cerr << ERROR_MSG_HEADER << " Failed to initialize TileDB.\n";
    return TILEDB_EINIT;
  }

  // Load binary file
  rc = tiledb_update_bin(
           tiledb_ctx, array_name.c_str(), path.c_str(), sorted);
  if(rc) {
    std::cerr << ERROR_MSG_HEADER << " Failed to load binary file(s).\n";
    return rc;
  }

  // Finalize TileDB
  rc = tiledb_finalize(tiledb_ctx);
  if(rc) {
    std::cerr << ERROR_MSG_HEADER << " Failed to finalize TileDB.\n";
    return TILEDB_EFIN;
  }

  std::cout << MSG_HEADER << " Program executed successfully!\n";
  
  return 0;
}
